gensim/models/poincare.py

```diff
@@ -0,0 +1,803 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Author: Jayant Jain <jayantjain1992@gmail.com>
+# Copyright (C) 2017 Radim Rehurek <me@radimrehurek.com>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+
+"""
+Python implementation of Poincare Embeddings, an embedding to capture hierarchical information,
+described in [1]
+
+This module allows training a Poincare Embedding from a training file containing relations from
+a transitive closure.
+
+
+.. [1] https://arxiv.org/pdf/1705.08039.pdf
+
+"""
+
+
+import csv
+import itertools
+import logging
+import os
+import random
+import time
+
+from autograd import numpy as np, grad
```

Comment:
Oh, new dependency for gensim. Why is needed (I mean why we can't differentiate loss analytically)?

Reply:
@jayantj Could we make it so that `autograd` is imported on demand only if `check_gradients` is requested? Looking at the code, this will require some refactoring (can't initialize `self.loss_grad` in the `__init__`), but it shouldn't be too hard to do.

It would also be nice to fail with a clear message if the import fails, such as "please `pip install autograd` if you want to check gradients" or something along those lines.

Then I suppose we can get away without adding `autograd` to gensim dependencies.

Reply:
IMO, checking gradients should definitely be enabled by default, a user is unlikely to enable it himself in case we disable it by default, therefore running the risk of not knowing in case something goes wrong with the training. In case we are enabling gradient checking by default, it makes sense to have `autograd` in the dependencies, since having the training fail due to an uninstalled dependency would be rather annoying.

If we are completely sure about the correctness of the model, we don't need the gradient checking in the first place - however, I imagine there are cases we haven't yet discovered where something might go wrong with the training. Would it be okay to add the dependency for now, and maybe once we are confident about correctness, we remove the gradient checking and the dependency?

Reply:
Yes, I agree with you. Variant with conditional import is good enough. Let's use `autograd`, but don't add it as a dependency (only in [test_env](https://github.com/RaRe-Technologies/gensim/blob/develop/setup.py#L228) of course).

Reply:
@jayantj Good points. The way I think about it is we shouldn't need gradient checking in the final version. Thus, it's OK to have it disabled by default. It's a useful feature for developers though, should there be any enhancements in the future, so I would definitely keep it in the code. The developers will know how to enable it.

That said, I believe what I have initially proposed would be ideal -- have the checking disabled by default; do the autograd import on demand; do not add another gensim dependency.

Reply:
Sounds good. Let's do conditional import. Maybe even instead of raising an error, just adding a prominent warning that gradient checking isn't happening, and to enable gradient check, the user should install autograd, would be good enough?

Reply:
Done https://github.com/RaRe-Technologies/gensim/pull/1696/commits/98e078dbc24e5cbadb08ed09077b6be561d48ccf

In case gradient checking is enabled and autograd isn't installed, I'm logging a warning that gradients checking isn't occurring and autograd needs to be installed to do so. Can raise an exception if that is better.

Reply:
@jayantj I think that exception isn't needed here (because model successfully works without `autograd`), the warning is enough.